File: server/internal/domain/upgrades.go
package domain

// TomeDef defines a Tome upgrade type.
type TomeDef struct {
	Type           TomeType `json:"type"`
	Name           string   `json:"name"`
	Description    string   `json:"description"`
	Tier           string   `json:"tier"`
	EffectPerStack float64  `json:"effectPerStack"`
	MaxStack       int      `json:"maxStack"`
}

// AbilityDef defines an Ability upgrade type.
type AbilityDef struct {
	Type            AbilityType `json:"type"`
	Name            string      `json:"name"`
	Description     string      `json:"description"`
	BaseCooldownSec float64     `json:"baseCooldownSec"` // 0 = passive
	AutoTrigger     string      `json:"autoTrigger"`
	BaseDamage      float64     `json:"baseDamage,omitempty"`
	Duration        float64     `json:"duration,omitempty"` // seconds
	UpgradeDmgMult  float64     `json:"upgradeDmgMult"`     // damage multiplier per level
	UpgradeCdMult   float64     `json:"upgradeCdMult"`      // cooldown multiplier per level
}

// SynergyDef defines a synergy bonus.
type SynergyDef struct {
	ID           string                `json:"id"`
	Name         string                `json:"name"`
	Description  string                `json:"description"`
	Hidden       bool                  `json:"hidden"`
	TomeReqs     map[TomeType]int      `json:"tomeReqs,omitempty"`
	AbilityReqs  map[AbilityType]int   `json:"abilityReqs,omitempty"` // min ability level
	BonusType    string                `json:"bonusType"`
	BonusValue   float64               `json:"bonusValue"`
}

// RequirementsMet checks if a PlayerBuild satisfies this synergy.
func (s *SynergyDef) RequirementsMet(build PlayerBuild) bool {
	for tome, required := range s.TomeReqs {
		if build.Tomes[tome] < required {
			return false
		}
	}
	for ability, requiredLevel := range s.AbilityReqs {
		found := false
		for _, slot := range build.AbilitySlots {
			if slot.Type == ability && slot.Level >= requiredLevel {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}

// --- XP Level Curve ---

// XPRequiredPerLevel defines XP needed to reach each level (index = level).
// Level 1 requires 0 XP (starting level).
var XPRequiredPerLevel = []int{
	0,   // Level 1 (start)
	20,  // Level 2
	30,  // Level 3
	45,  // Level 4
	65,  // Level 5
	90,  // Level 6
	120, // Level 7
	155, // Level 8
	195, // Level 9
	240, // Level 10
	290, // Level 11
	345, // Level 12
}

// MaxLevel is the maximum agent level.
const MaxLevel = 12

// XPForLevel returns the XP required to reach a given level (1-indexed).
// Returns -1 if beyond max level.
func XPForLevel(level int) int {
	if level < 1 || level > MaxLevel {
		return -1
	}
	return XPRequiredPerLevel[level-1]
}

// CumulativeXPForLevel returns total XP needed from level 1 to reach this level.
func CumulativeXPForLevel(level int) int {
	if level < 1 || level > MaxLevel {
		return -1
	}
	total := 0
	for i := 0; i < level; i++ {
		total += XPRequiredPerLevel[i]
	}
	return total
}

// --- Tome Definitions (8 types) ---

// AllTomes defines all 8 Tome upgrades.
var AllTomes = []TomeDef{
	{Type: TomeXP, Name: "XP Tome", Description: "XP gain +20% per stack", Tier: "S", EffectPerStack: 0.20, MaxStack: 10},
	{Type: TomeSpeed, Name: "Speed Tome", Description: "Move speed +10% per stack", Tier: "S", EffectPerStack: 0.10, MaxStack: 5},
	{Type: TomeDamage, Name: "Damage Tome", Description: "Aura DPS +15% per stack", Tier: "S", EffectPerStack: 0.15, MaxStack: 10},
	{Type: TomeArmor, Name: "Armor Tome", Description: "Damage taken -10% per stack", Tier: "A", EffectPerStack: 0.10, MaxStack: 8},
	{Type: TomeMagnet, Name: "Magnet Tome", Description: "Orb collect radius +25% per stack", Tier: "A", EffectPerStack: 0.25, MaxStack: 6},
	{Type: TomeLuck, Name: "Luck Tome", Description: "Rare upgrade chance +15% per stack", Tier: "A", EffectPerStack: 0.15, MaxStack: 6},
	{Type: TomeRegen, Name: "Regen Tome", Description: "Regenerate +0.5 mass/s per stack", Tier: "B", EffectPerStack: 0.025, MaxStack: 5}, // 0.025 mass/tick = 0.5 mass/s at 20Hz
	{Type: TomeCursed, Name: "Cursed Tome", Description: "DPS +25%, damage taken +20% per stack", Tier: "S*", EffectPerStack: 0.25, MaxStack: 5},
}

// GetTomeDef returns the TomeDef for a given TomeType, or nil if not found.
func GetTomeDef(t TomeType) *TomeDef {
	for i := range AllTomes {
		if AllTomes[i].Type == t {
			return &AllTomes[i]
		}
	}
	return nil
}

// --- Ability Definitions (6 types) ---

// AllAbilities defines all 6 Ability upgrades.
var AllAbilities = []AbilityDef{
	{Type: AbilityVenomAura, Name: "Venom Aura", Description: "Poison DoT to nearby enemies (3s)",
		BaseCooldownSec: 0, AutoTrigger: "always_active", BaseDamage: 15, Duration: 3,
		UpgradeDmgMult: 1.30, UpgradeCdMult: 1.0},
	{Type: AbilityShieldBurst, Name: "Shield Burst", Description: "3s invincibility + knockback",
		BaseCooldownSec: 15, AutoTrigger: "hp_below_30pct", BaseDamage: 0, Duration: 3,
		UpgradeDmgMult: 1.0, UpgradeCdMult: 0.80},
	{Type: AbilityLightningStrike, Name: "Lightning Strike", Description: "Instant damage to nearest enemy",
		BaseCooldownSec: 8, AutoTrigger: "enemy_in_range", BaseDamage: 50, Duration: 0,
		UpgradeDmgMult: 1.30, UpgradeCdMult: 0.80},
	{Type: AbilitySpeedDash, Name: "Speed Dash", Description: "2s triple speed + collision immunity",
		BaseCooldownSec: 12, AutoTrigger: "chase_or_flee", BaseDamage: 0, Duration: 2,
		UpgradeDmgMult: 1.0, UpgradeCdMult: 0.80},
	{Type: AbilityMassDrain, Name: "Mass Drain", Description: "Absorb 10% mass on contact",
		BaseCooldownSec: 10, AutoTrigger: "on_contact", BaseDamage: 0, Duration: 0,
		UpgradeDmgMult: 1.30, UpgradeCdMult: 0.80},
	{Type: AbilityGravityWell, Name: "Gravity Well", Description: "3s pull nearby orbs + enemies",
		BaseCooldownSec: 20, AutoTrigger: "orb_dense_area", BaseDamage: 0, Duration: 3,
		UpgradeDmgMult: 1.0, UpgradeCdMult: 0.80},
}

// GetAbilityDef returns the AbilityDef for a given AbilityType, or nil if not found.
func GetAbilityDef(a AbilityType) *AbilityDef {
	for i := range AllAbilities {
		if AllAbilities[i].Type == a {
			return &AllAbilities[i]
		}
	}
	return nil
}

// --- Synergy Definitions (6 public + 4 hidden = 10) ---

// AllSynergies defines all 10 synergy bonuses.
var AllSynergies = []SynergyDef{
	// === Public Synergies (6) ===
	{
		ID: "holy_trinity", Name: "Holy Trinity",
		Description: "All XP +50% bonus",
		Hidden: false,
		TomeReqs:    map[TomeType]int{TomeXP: 3, TomeLuck: 2, TomeCursed: 1},
		BonusType:   "xp_multiplier",
		BonusValue:  0.50,
	},
	{
		ID: "glass_cannon", Name: "Glass Cannon",
		Description: "DPS x2.0 (damage taken x2.0)",
		Hidden: false,
		TomeReqs:    map[TomeType]int{TomeDamage: 5, TomeCursed: 3},
		BonusType:   "dps_multiplier",
		BonusValue:  2.0,
	},
	{
		ID: "iron_fortress", Name: "Iron Fortress",
		Description: "Additional -30% damage taken",
		Hidden: false,
		TomeReqs:    map[TomeType]int{TomeArmor: 4, TomeRegen: 3},
		AbilityReqs: map[AbilityType]int{AbilityShieldBurst: 1},
		BonusType:   "damage_reduction",
		BonusValue:  0.30,
	},
	{
		ID: "speedster", Name: "Speedster",
		Description: "Boost cost -50%",
		Hidden: false,
		TomeReqs:    map[TomeType]int{TomeSpeed: 4, TomeMagnet: 2},
		AbilityReqs: map[AbilityType]int{AbilitySpeedDash: 1},
		BonusType:   "boost_cost_reduction",
		BonusValue:  0.50,
	},
	{
		ID: "vampire", Name: "Vampire",
		Description: "Poison damage heals 20% as mass",
		Hidden: false,
		TomeReqs:    map[TomeType]int{TomeRegen: 2},
		AbilityReqs: map[AbilityType]int{AbilityVenomAura: 1, AbilityMassDrain: 1},
		BonusType:   "poison_lifesteal",
		BonusValue:  0.20,
	},
	{
		ID: "storm", Name: "Storm",
		Description: "Lightning chains to 3 enemies",
		Hidden: false,
		TomeReqs:    map[TomeType]int{TomeDamage: 3},
		AbilityReqs: map[AbilityType]int{AbilityLightningStrike: 3},
		BonusType:   "lightning_chain",
		BonusValue:  3.0,
	},

	// === Hidden Synergies (4) ===
	{
		ID: "berserker", Name: "Berserker",
		Description: "Dash damage x3.0",
		Hidden: true,
		TomeReqs:    map[TomeType]int{TomeDamage: 4, TomeCursed: 2, TomeSpeed: 2},
		BonusType:   "dash_damage_multiplier",
		BonusValue:  3.0,
	},
	{
		ID: "arcane_scholar", Name: "Arcane Scholar",
		Description: "Tome effects +25% globally",
		Hidden: true,
		TomeReqs:    map[TomeType]int{TomeXP: 4, TomeLuck: 4, TomeMagnet: 3},
		BonusType:   "tome_effect_multiplier",
		BonusValue:  0.25,
	},
	{
		ID: "immortal", Name: "Immortal",
		Description: "Survive lethal hit once per round (1 HP)",
		Hidden: true,
		TomeReqs:    map[TomeType]int{TomeArmor: 6, TomeRegen: 4},
		AbilityReqs: map[AbilityType]int{AbilityShieldBurst: 2},
		BonusType:   "cheat_death",
		BonusValue:  1.0,
	},
	{
		ID: "gravity_master", Name: "Gravity Master",
		Description: "Gravity Well radius x2 + permanent slow aura",
		Hidden: true,
		TomeReqs:    map[TomeType]int{TomeMagnet: 5, TomeSpeed: 3},
		AbilityReqs: map[AbilityType]int{AbilityGravityWell: 2},
		BonusType:   "gravity_enhance",
		BonusValue:  2.0,
	},
}

// GetSynergyDef returns the SynergyDef with the given ID, or nil if not found.
func GetSynergyDef(id string) *SynergyDef {
	for i := range AllSynergies {
		if AllSynergies[i].ID == id {
			return &AllSynergies[i]
		}
	}
	return nil
}

// PublicSynergies returns only the non-hidden synergies.
func PublicSynergies() []SynergyDef {
	var result []SynergyDef
	for _, s := range AllSynergies {
		if !s.Hidden {
			result = append(result, s)
		}
	}
	return result
}

File: packages/shared/src/constants/upgrades.ts
/**
 * Agent Survivor v10 — Upgrade System Constants
 * Tomes, Abilities, Synergies, XP Table, Combat Config
 */

import type { TomeType, AbilityType, SynergyDef, MapObjectType, MapObjectConfig } from '../types/game';

// ─── XP Table (Level → Required XP) ───

/** 레벨업에 필요한 XP (인덱스 = 현재 레벨, 값 = 다음 레벨까지 필요 XP) */
export const XP_TABLE: readonly number[] = [
  0,    // Lv0→1: 시작
  20,   // Lv1→2
  30,   // Lv2→3
  45,   // Lv3→4
  65,   // Lv4→5
  90,   // Lv5→6
  120,  // Lv6→7
  155,  // Lv7→8
  195,  // Lv8→9
  240,  // Lv9→10
  290,  // Lv10→11
  345,  // Lv11→12
  999999, // Lv12 (최대 레벨)
] as const;

export const MAX_LEVEL = 12;

// ─── XP Sources ───

export const XP_SOURCE = {
  NATURAL_ORB: 1,       // 자연 오브 수집
  DEATH_ORB: 4,         // 사망 오브 수집
  POWER_UP_ORB: 5,      // 파워업 오브 수집
  AURA_KILL_BASE: 10,   // 오라 전투 킬 기본
  AURA_KILL_PER_LEVEL: 3, // + 적 레벨 × 3
  DASH_KILL_BASE: 15,   // 대시 킬 기본
  DASH_KILL_PER_LEVEL: 3, // + 적 레벨 × 3
} as const;

export const KILL_STREAK_MULTIPLIER = {
  3: 1.5,   // 3연속 킬
  5: 2.0,   // 5연속 킬
  10: 3.0,  // 10연속 킬
} as const;

// ─── Tome Definitions ───

export interface TomeDef {
  type: TomeType;
  name: string;
  tier: 'S' | 'A' | 'B';
  effectPerStack: number;  // 스택당 효과 (%)
  maxStacks: number;
  description: string;
  statKey: string;  // 적용되는 스탯 키
}

export const TOME_DEFS: Record<TomeType, TomeDef> = {
  xp: {
    type: 'xp',
    name: 'XP Tome',
    tier: 'S',
    effectPerStack: 0.20,  // +20% XP per stack
    maxStacks: 10,
    description: 'XP 획득 +20%',
    statKey: 'xpMultiplier',
  },
  speed: {
    type: 'speed',
    name: 'Speed Tome',
    tier: 'S',
    effectPerStack: 0.10,  // +10% speed per stack
    maxStacks: 5,
    description: '이동속도 +10%',
    statKey: 'speedMultiplier',
  },
  damage: {
    type: 'damage',
    name: 'Damage Tome',
    tier: 'S',
    effectPerStack: 0.15,  // +15% DPS per stack
    maxStacks: 10,
    description: '오라 DPS +15%',
    statKey: 'dpsMultiplier',
  },
  armor: {
    type: 'armor',
    name: 'Armor Tome',
    tier: 'A',
    effectPerStack: 0.10,  // -10% damage taken per stack
    maxStacks: 8,
    description: '받는 데미지 -10%',
    statKey: 'damageReduction',
  },
  magnet: {
    type: 'magnet',
    name: 'Magnet Tome',
    tier: 'A',
    effectPerStack: 0.25,  // +25% collect radius per stack
    maxStacks: 6,
    description: '오브 수집 반경 +25%',
    statKey: 'collectRadiusMultiplier',
  },
  luck: {
    type: 'luck',
    name: 'Luck Tome',
    tier: 'A',
    effectPerStack: 0.15,  // +15% rare upgrade chance per stack
    maxStacks: 6,
    description: '레어 업그레이드 확률 +15%',
    statKey: 'luckMultiplier',
  },
  regen: {
    type: 'regen',
    name: 'Regen Tome',
    tier: 'B',
    effectPerStack: 0.025, // +0.025 mass/tick (=0.5/s) per stack
    maxStacks: 5,
    description: '매초 mass +0.5 회복',
    statKey: 'regenPerTick',
  },
  cursed: {
    type: 'cursed',
    name: 'Cursed Tome',
    tier: 'S',
    effectPerStack: 0.25,  // +25% DPS, +20% damage taken per stack
    maxStacks: 5,
    description: 'DPS +25%, 받는 피해 +20%',
    statKey: 'cursedMultiplier',
  },
} as const;

// ─── Ability Definitions ───

export interface AbilityDef {
  type: AbilityType;
  name: string;
  baseDamage: number;       // 기본 피해량 (0 = 비전투)
  cooldownTicks: number;    // 쿨다운 (틱 단위)
  durationTicks: number;    // 지속 시간 (0 = 즉시)
  autoTrigger: string;      // 자동 발동 조건 설명
  description: string;
  upgradeBonus: {
    damageMultiplier: number;   // 강화당 피해 배율
    cooldownReduction: number;  // 강화당 쿨다운 감소율
  };
}

export const ABILITY_DEFS: Record<AbilityType, AbilityDef> = {
  venom_aura: {
    type: 'venom_aura',
    name: 'Venom Aura',
    baseDamage: 15,         // 15 mass/s DoT
    cooldownTicks: 0,       // 패시브
    durationTicks: 60,      // 독 3초
    autoTrigger: '항상 활성',
    description: '근접 적에게 독 DoT (3s)',
    upgradeBonus: { damageMultiplier: 1.3, cooldownReduction: 0 },
  },
  shield_burst: {
    type: 'shield_burst',
    name: 'Shield Burst',
    baseDamage: 0,
    cooldownTicks: 300,     // 15초
    durationTicks: 60,      // 3초 무적
    autoTrigger: '체력 30% 이하',
    description: '3초간 무적 + 주변 적 밀침',
    upgradeBonus: { damageMultiplier: 1, cooldownReduction: 0.20 },
  },
  lightning_strike: {
    type: 'lightning_strike',
    name: 'Lightning Strike',
    baseDamage: 50,
    cooldownTicks: 160,     // 8초
    durationTicks: 0,       // 즉시
    autoTrigger: '사거리 내 적 감지',
    description: '가장 가까운 적에게 순간 데미지',
    upgradeBonus: { damageMultiplier: 1.3, cooldownReduction: 0.20 },
  },
  speed_dash: {
    type: 'speed_dash',
    name: 'Speed Dash',
    baseDamage: 0,
    cooldownTicks: 240,     // 12초
    durationTicks: 40,      // 2초
    autoTrigger: '적 추격/도주 시',
    description: '2초간 3배 속도 + 충돌 면역',
    upgradeBonus: { damageMultiplier: 1, cooldownReduction: 0.20 },
  },
  mass_drain: {
    type: 'mass_drain',
    name: 'Mass Drain',
    baseDamage: 0,          // 비례 데미지 (mass의 10%)
    cooldownTicks: 200,     // 10초
    durationTicks: 0,       // 즉시
    autoTrigger: '히트박스 충돌 시',
    description: '터치한 적의 mass 10% 흡수',
    upgradeBonus: { damageMultiplier: 1.3, cooldownReduction: 0.20 },
  },
  gravity_well: {
    type: 'gravity_well',
    name: 'Gravity Well',
    baseDamage: 0,
    cooldownTicks: 400,     // 20초
    durationTicks: 60,      // 3초
    autoTrigger: '오브 밀집 지역',
    description: '3초간 주변 오브+적을 끌어당김',
    upgradeBonus: { damageMultiplier: 1, cooldownReduction: 0.20 },
  },
} as const;

// ─── Synergy Definitions ───

export const ALL_SYNERGIES: SynergyDef[] = [
  // 공개 시너지 (6종)
  {
    id: 'holy_trinity',
    name: 'Holy Trinity',
    requirements: { tomes: { xp: 3, luck: 2, cursed: 1 } },
    bonus: { description: '모든 XP +50% 추가', effects: { xpMultiplier: 1.5 } },
    hidden: false,
  },
  {
    id: 'glass_cannon',
    name: 'Glass Cannon',
    requirements: { tomes: { damage: 5, cursed: 3 } },
    bonus: { description: 'DPS ×2.0 (피해도 ×2.0)', effects: { dpsMultiplier: 2.0, damageTakenMultiplier: 2.0 } },
    hidden: false,
  },
  {
    id: 'iron_fortress',
    name: 'Iron Fortress',
    requirements: { tomes: { armor: 4, regen: 3 }, abilities: { shield_burst: 1 } },
    bonus: { description: '받는 피해 추가 -30%', effects: { damageReduction: 0.30 } },
    hidden: false,
  },
  {
    id: 'speedster',
    name: 'Speedster',
    requirements: { tomes: { speed: 4, magnet: 2 }, abilities: { speed_dash: 1 } },
    bonus: { description: '부스트 비용 -50%', effects: { boostCostReduction: 0.50 } },
    hidden: false,
  },
  {
    id: 'vampire',
    name: 'Vampire',
    requirements: { tomes: { regen: 2 }, abilities: { venom_aura: 1, mass_drain: 1 } },
    bonus: { description: '독 피해의 20%를 mass 회복', effects: { venomLifesteal: 0.20 } },
    hidden: false,
  },
  {
    id: 'storm',
    name: 'Storm',
    requirements: { tomes: { damage: 3 }, abilities: { lightning_strike: 3 } },
    bonus: { description: '번개 연쇄 (3 적까지)', effects: { lightningChain: 3 } },
    hidden: false,
  },
  // 히든 시너지 (4종)
  {
    id: 'berserker',
    name: '???',
    requirements: { tomes: { cursed: 5, damage: 3, speed: 2 } },
    bonus: { description: '대시 피해 ×3.0', effects: { dashDamageMultiplier: 3.0 } },
    hidden: true,
  },
  {
    id: 'pacifist',
    name: '???',
    requirements: { tomes: { xp: 5, magnet: 4, regen: 3 } },
    bonus: { description: 'XP 오브 가치 ×3.0', effects: { orbValueMultiplier: 3.0 } },
    hidden: true,
  },
  {
    id: 'elemental',
    name: '???',
    requirements: { abilities: { venom_aura: 2, lightning_strike: 2, gravity_well: 1 } },
    bonus: { description: '모든 Ability 쿨다운 -40%', effects: { abilityCooldownReduction: 0.40 } },
    hidden: true,
  },
  {
    id: 'immortal',
    name: '???',
    requirements: { tomes: { armor: 6, regen: 4 }, abilities: { shield_burst: 3 } },
    bonus: { description: '사망 시 1회 부활 (mass 50%)', effects: { reviveOnce: 1, reviveMassRatio: 0.50 } },
    hidden: true,
  },
];

// ─── Combat Config ───

export const COMBAT_CONFIG = {
  AURA_RADIUS: 60,                // 전투 오라 반경 (px)
  BASE_AURA_DPS_PER_TICK: 2.0,   // 기본 오라 DPS (mass/tick)
  DASH_DAMAGE_RATIO: 0.30,        // 대시 충돌 시 상대 mass의 30% 즉시 피해
  HITBOX_BASE_RADIUS: 16,         // 기본 히트박스 반경 (px)
  HITBOX_MAX_RADIUS: 22,          // 최대 히트박스 반경 (px)
  HIGH_LEVEL_THRESHOLD: 8,        // 고레벨 기준
  HIGH_LEVEL_DPS_BONUS: 0.20,     // 고레벨 DPS 보너스 (+20%)
  VENOM_DPS_PER_TICK: 0.75,       // 독 DoT (15 mass/s ÷ 20Hz)
  VENOM_DURATION_TICKS: 60,       // 독 지속 3초
  MASS_DRAIN_RATIO: 0.10,         // Mass Drain: 10% 흡수
  GRAVITY_PULL_RADIUS: 200,       // Gravity Well 끌어당김 반경
  GRAVITY_PULL_SPEED: 5,          // Gravity Well 끌어당김 속도
  SHIELD_PUSH_RADIUS: 80,         // Shield Burst 밀침 반경
  SHIELD_PUSH_FORCE: 50,          // Shield Burst 밀침 거리
  LIGHTNING_RANGE: 200,           // Lightning Strike 사거리
  DEATH_XP_ORB_RATIO: 0.80,      // 사망 시 mass의 80%가 XP 오브
} as const;

// ─── Arena Shrink Config ───

export const SHRINK_CONFIG = {
  ENABLED: true,
  INITIAL_RADIUS: 6000,           // 초기 반경
  MIN_RADIUS: 1200,               // 최소 반경
  SHRINK_RATE_PER_MIN: 600,       // 분당 수축 (px)
  /** 틱당 수축 = 600/60/20 = 0.5 px/tick */
  SHRINK_PER_TICK: 0.5,
  BOUNDARY_PENALTY_PER_TICK: 0.0025, // 경계 밖 패널티 (mass의 0.25%/tick)
} as const;

// ─── Grace Period ───

export const GRACE_PERIOD_TICKS = 600; // 30초 @ 20Hz — 초반 전투 면역

// ─── Upgrade Choice Config ───

export const UPGRADE_CONFIG = {
  CHOICES_PER_LEVEL: 3,            // 레벨업 시 3택
  CHOICE_TIMEOUT_TICKS: 100,       // 5초 타임아웃
  MAX_ABILITY_SLOTS: 2,            // 기본 Ability 슬롯 2개
  MAX_ABILITY_LEVEL: 4,            // Ability 최대 강화 4회
  /** Ability가 나올 확률 (나머지는 Tome) */
  ABILITY_OFFER_CHANCE: 0.35,
  /** Luck Tome 스택당 레어 업그레이드 확률 증가 */
  LUCK_RARE_BONUS_PER_STACK: 0.15,
} as const;

// ─── Bot Build Paths ───

export type BotBuildPath = 'aggressive' | 'tank' | 'xp_rush' | 'balanced' | 'glass_cannon';

export const BOT_BUILD_PREFERENCES: Record<BotBuildPath, {
  tomePriority: TomeType[];
  abilityPriority: AbilityType[];
}> = {
  aggressive: {
    tomePriority: ['damage', 'speed', 'cursed'],
    abilityPriority: ['venom_aura', 'lightning_strike', 'speed_dash'],
  },
  tank: {
    tomePriority: ['armor', 'regen', 'magnet'],
    abilityPriority: ['shield_burst', 'mass_drain', 'gravity_well'],
  },
  xp_rush: {
    tomePriority: ['xp', 'magnet', 'speed'],
    abilityPriority: ['gravity_well', 'speed_dash', 'venom_aura'],
  },
  balanced: {
    tomePriority: ['damage', 'armor', 'speed'],
    abilityPriority: ['venom_aura', 'shield_burst', 'lightning_strike'],
  },
  glass_cannon: {
    tomePriority: ['damage', 'cursed', 'speed'],
    abilityPriority: ['lightning_strike', 'venom_aura', 'speed_dash'],
  },
};

// ─── Map Object Config ───

export const MAP_OBJECT_CONFIGS: Record<MapObjectType, MapObjectConfig> = {
  shrine: {
    type: 'shrine',
    name: 'XP Shrine',
    radius: 48,
    respawnTicks: 600,     // 30초 @ 20Hz
    count: 3,
    description: '+15 XP on interact',
  },
  spring: {
    type: 'spring',
    name: 'Healing Spring',
    radius: 48,
    respawnTicks: 0,        // 항상 활성
    count: 2,
    description: '+5 mass/tick while in range',
  },
  altar: {
    type: 'altar',
    name: 'Upgrade Altar',
    radius: 48,
    respawnTicks: 1200,    // 60초 @ 20Hz
    count: 1,
    description: 'Instant level-up (1 use)',
  },
  gate: {
    type: 'gate',
    name: 'Speed Gate',
    radius: 32,
    respawnTicks: 200,     // 10초 @ 20Hz
    count: 2,
    description: '3s speed boost on pass',
  },
};
